### PR TITLE
rgw: using corresponding index shard for objects in RGWRados::remove_…

### PR DESCRIPTION
**Pull request by shashalu**
…objs_from_index

Fixes: http://tracker.ceph.com/issues/20355

Signed-off-by: Shasha Lu <lu.shasha@eisoo.com>

### CODE REVIEW

**Comment by oritwas** [comment]
@shashalu , I suggest using RGWRados::BucketShard this will remove the need for parse_hash_object_from_indexkey

**Comment by shashalu** [comment]
@oritwas 
'rgw_obj_index_key' only has name and instance.  To get objects shard_id, we must get hash_object first. 
For ordinary objects, name == hash_object
For multipart objects name = _multipart_ObjectName.uuid.partnum, but hash_object = ObjectName.

RGWRados::BucketShard::init(const rgw_bucket& _bucket, const rgw_obj& obj); The rgw_obj contains index_hash_source. For ordinary objects the value is empty, whe get_hash_object() will retrun object name. For multipart objects, the value is origin object name which is not the same with multipart object name. 

Without parse_hash_object_from_indexkey, we cannot got the correctly hash_object.

**Comment by oritwas** [comment]
You can change remove_objs_from_index to get a list of rgw_obj instead of rgw_obj_index_key

**Comment by shashalu** [comment]
ping @oritwas

File: src/rgw/rgw_bucket.cc [hunk 1]
```diff
@@ -1056,7 +1056,7 @@ int RGWBucket::check_bad_index_multipart(RGWBucketAdminOpState& op_state,
       int pos = oid.find_last_of('.');
       if (pos < 0) {
         /* obj has no suffix */
-        all_objs[key] = oid;
```
Comment (oritwas) [inline src/rgw/rgw_bucket.cc]: you can remove the key variable and change rgw_obj obj(bucket, key) to  rgw_obj obj(bucket,  iter->key)

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -8470,24 +8470,6 @@ int RGWRados::send_chain_to_gc(cls_rgw_obj_chain& chain, const string& tag, bool
   return gc->send_chain(chain, tag, sync);
 }
 
-int RGWRados::open_bucket_index(const RGWBucketInfo& bucket_info, librados::IoCtx& index_ctx, string& bucket_oid)
```
Comment (oritwas) [inline src/rgw/rgw_rados.cc]: please leave the code, we may use it someday

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -12868,33 +12850,30 @@ int RGWRados::cls_obj_usage_log_trim(string& oid, string& user, uint64_t start_e
...
   for (auto iter = oid_list.begin(); iter != oid_list.end(); ++iter) {
+    bufferlist updates, out;
     rgw_bucket_dir_entry entry;
-    entry.key = *iter;
+    iter->key.get_index_key(&entry.key);
     dout(2) << "RGWRados::remove_objs_from_index bucket=" << bucket_info.bucket << " obj=" << entry.key.name << ":" << entry.key.instance << dendl;
     entry.ver.epoch = (uint64_t)-1; // ULLONG_MAX, needed to that objclass doesn't skip out request
     updates.append(CEPH_RGW_REMOVE | suggest_flag);
     ::encode(entry, updates);
+    RGWRados::BucketShard bs(this);
```
Comment (oritwas) [inline src/rgw/rgw_rados.cc]: now we are running dir_suggest one object at a time this won't scale when we have lots of objects.
remove_objs_from_index is called in two cases:
1. radosgw-admin obj unlink - that is used for a single object and than there is no scaling issue (and which fails)
2. check_bad_index_multipart (called when listing a bucket and when running bucket check --fix) here the scale is a major factor.

The only solution I see is using a different function for radosgw-admin obj unlink.
You will need to add a new method remove_obj_from_index that gets a single object and removes it (using the BucketShard).

Comment (shashalu) [reply src/rgw/rgw_rados.cc]: @oritwas  I think there is no need to add a new method, radosgw-admin obj unlink can use this method too.  In radosgw-admin obj unlink the oid_list size is always 1.

Comment (oritwas) [reply src/rgw/rgw_rados.cc]: @shashalu, the problem is the changes effect object list and bucket check commands. those needs the performance and scale as they call the methods with a long list of objects

**Pull request closed without merging**